File: object-c/08_stack/LinkedStack.m
//
//  LinkedStack.m
//  algo
//
//  Created by Wenru Dong on 2018/10/8.
//  Copyright © 2018年 Wenru Dong. All rights reserved.
//

#import "LinkedStack.h"
#import "ListNode.h"

@implementation LinkedStack
{
    @private
    ListNode* _top;
}

- (BOOL)isEmpty {
    return _top == nil;
}

- (void)push:(int)value {
    ListNode *newTop = [ListNode nodeWithValue:value];
    newTop.next = _top;
    _top = newTop;
}

- (int)pop {
    if ([self isEmpty]) {
        [NSException raise:NSRangeException format:@"The stack is empty."];
    }
    int value = _top.value;
    _top = _top.next;
    return value;
}

- (NSString *)debugDescription {
    NSMutableString *info = [[NSMutableString alloc] init];
    ListNode *current = _top;
    while (current) {
        [info appendString:[NSString stringWithFormat:@"%d]", current.value]];
        current = current.next;
    }
    return [NSString stringWithString:info];
}

@end

File: csharp/08-stack/algo08_stack/LinkedStack.cs
using System;

namespace algo08_stack
{
    public class LinkedStack<T>
    {
        private StackListNode<T> _top;

        public int Count { get; private set; }

        public void Push(T val)
        {
            var newNode = new StackListNode<T>(val);
            newNode.Next = _top;
            _top = newNode;

            Count++;
        }

        public T Pop()
        {
            if (_top == null) throw new InvalidOperationException("Stack empty");

            T val = _top.Value;
            _top = _top.Next;

            Count--;

            return val;
        }

        public void Clear()
        {
            while (Count > 0)
            {
                Pop();
            }
        }
    }

    public class StackListNode<T>
    {
        public StackListNode(T nodeValue)
        {
            Value = nodeValue;
        }

        public T Value { get; set; }
        public StackListNode<T> Next { get; set; }
    }
}